### PR TITLE
process id reuse policy

### PR DESCRIPTION
**Pull request by duoertai**

### CODE REVIEW

File: persistence/sql/sql_process_store.go [hunk 1]
```diff
@@ -62,20 +63,92 @@ func (p sqlProcessStoreImpl) doStartProcessTx(
 	prcExeId := uuid.MustNewUUID()
 	hasNewWorkerTask := false
 
-	err := tx.InsertCurrentProcessExecution(ctx, extensions.CurrentProcessExecutionRow{
+	if req.ProcessStartConfig == nil {
```
Comment (longquanzheng) [inline persistence/sql/sql_process_store.go:66]: I don't feel like it's necessary to require this field. All the configs inside are "optional".

File: persistence/sql/sql_process_store.go [hunk 2]
```diff
@@ -62,20 +63,92 @@ func (p sqlProcessStoreImpl) doStartProcessTx(
...
+		return nil, fmt.Errorf("processStartConfig is required")
+	}
+
+	prevProcessExecution, errGetPrevExecution := tx.SelectProcessExecutionForUpdate(ctx, prcExeId)
```
Comment (longquanzheng) [inline persistence/sql/sql_process_store.go:70]: This is a new executionId, there is no execution existing yet for this id.

Comment (longquanzheng) [reply persistence/sql/sql_process_store.go:70]: looks like you wanted to get the prevExecution, it should be after selectCurrentExecution

File: persistence/sql/sql_process_store.go [hunk 3]
```diff
@@ -62,20 +63,92 @@ func (p sqlProcessStoreImpl) doStartProcessTx(
...
-	if err != nil {
-		if p.session.IsDupEntryError(err) {
-			// TODO support other ProcessIdReusePolicy on this error
-			return &persistence.StartProcessResponse{
-				AlreadyStarted: true,
-			}, nil
+	if errInsertCurrentProcessExecution != nil {
+		if p.session.IsDupEntryError(errInsertCurrentProcessExecution) {
+			// process with the same id is running
+			if req.ProcessStartConfig.GetIdReusePolicy() == xdbapi.TERMINATE_IF_RUNNING {
```
Comment (longquanzheng) [inline persistence/sql/sql_process_store.go:81]: it's better to put each policy handling in a smaller method

File: persistence/sql/sql_process_store.go (hunk 1 above)
Comment (duoertai) [reply persistence/sql/sql_process_store.go:66]: default to reject_duplicate?

Comment (longquanzheng) [reply persistence/sql/sql_process_store.go:66]: In cadence the default is “failed only”(if previous exit abnormally). Then Temporal changed it to “allow if not running “. The reason behind that is that new users would just ask why their workflow cannot be started. It seems more friendly to new users. I would trust the decision from Temporal team for this default behavior

File: persistence/sql/sql_process_store.go (hunk 2 above)
Comment (duoertai) [reply persistence/sql/sql_process_store.go:70]: if there is a running process with the same id, and id reuse policy is terminate_if_running, we still need the previous execution, right?

Comment (longquanzheng) [reply persistence/sql/sql_process_store.go:70]: Yes, to get the previous execution, we need to get the processExeucitonID from selectCurrentExecution. Doing it before selectCurrentExecution will have no value? 

Note that the processExeucitonID is uniquely created for every request.

File: extensions/tests/base_sql.go [hunk 4]
```diff
@@ -0,0 +1,278 @@
+// Apache License 2.0
```
Comment (longquanzheng) [inline extensions/tests/base_sql.go:1]: I have refactored the test to https://github.com/xdblab/xdb/blob/main/persistence/sql/basetest to make it easier to maintain

File: extensions/postgres/transactional.go [hunk 5]
```diff
@@ -5,27 +5,46 @@
...
 
-func (d dbTx) InsertCurrentProcessExecution(ctx context.Context, row extensions.CurrentProcessExecutionRow) error {
+func (d dbTx) InsertLatestProcessExecution(ctx context.Context, row extensions.LatestProcessExecutionRow) error {
 	row.ProcessExecutionIdString = row.ProcessExecutionId.String()
-	_, err := d.tx.ExecContext(ctx, insertCurrentProcessExecutionQuery, row.Namespace, row.ProcessId, row.ProcessExecutionIdString)
+	_, err := d.tx.ExecContext(ctx, insertLatestProcessExecutionQuery, row.Namespace, row.ProcessId, row.ProcessExecutionIdString)
+	return err
+}
+
+const selectLatestProcessExecutionForUpdateQuery = `SELECT namespace, process_id, process_execution_id FROM xdb_sys_latest_process_executions WHERE namespace=$1 AND process_id=$2 FOR UPDATE`
```
Comment (longquanzheng) [inline extensions/postgres/transactional.go:34]: For any new query/SQL interface, we should test in https://github.com/xdblab/xdb/blob/main/persistence/sql/basetest/base_basic.go 
(see my similar comment on https://github.com/xdblab/xdb/pull/44#discussion_r1359433331 )

File: engine/worker_task_conrrent_processor.go [hunk 6]
```diff
@@ -186,7 +187,7 @@ func (w *workerTaskConcurrentProcessor) processWaitUntilTask(
 	}
 	if checkHttpError(err, httpResp) {
 		err := composeHttpError(err, httpResp)
-		w.logger.Info("worker API return error", tag.Error(err))
+		//w.logger.Info("worker API return error", tag.Error(err))
```
Comment (longquanzheng) [inline engine/worker_task_conrrent_processor.go:190]: curious why this is commented out?

Comment (duoertai) [reply engine/worker_task_conrrent_processor.go:190]: just for debug, sometimes there are too many logs so that I can't find the error log, will revert this change

File: extensions/postgres/transactional.go [hunk 7]
```diff
@@ -15,17 +15,37 @@ package postgres
...
+
+const selectLatestProcessExecutionForUpdateQuery = `SELECT namespace, process_id, process_execution_id 
+FROM xdb_sys_latest_process_executions 
+WHERE namespace=$1 AND process_id=$2 FOR UPDATE`
+
+func (d dbTx) SelectLatestProcessExecutionForUpdate(
+	ctx context.Context, namespace string, processId string,
+) ([]extensions.LatestProcessExecutionRow, error) {
+	var rows []extensions.LatestProcessExecutionRow
+	err := d.tx.SelectContext(ctx, &rows, selectLatestProcessExecutionForUpdateQuery, namespace, processId)
```
Comment (longquanzheng) [inline extensions/postgres/transactional.go:41]: If it's expecting a single row, then use GetContext to ensure only single row is returned. https://jmoiron.github.io/sqlx/

File: persistence/sql/sql_process_store.go [hunk 8]
```diff
@@ -91,28 +92,372 @@ func (p sqlProcessStoreImpl) StopProcess(ctx context.Context, request persistenc
 	return resp, err
 }
 
+func (p sqlProcessStoreImpl) markPendingStateAsAborted(
```
Comment (longquanzheng) [inline persistence/sql/sql_process_store.go:95]: After rebase, you will see that @zklgame has created a new way to mark the states aborted in a batch query, which is more efficient than the way that I did

File: persistence/sql/sql_process_store.go [hunk 9]
```diff
@@ -91,28 +92,372 @@ func (p sqlProcessStoreImpl) StopProcess(ctx context.Context, request persistenc
...
 	req := request.Request
+
+	if req.ProcessStartConfig == nil {
+		req.ProcessStartConfig = &xdbapi.ProcessStartConfig{
+			IdReusePolicy: xdbapi.ALLOW_IF_NO_RUNNING.Ptr(),
+		}
+	}
+
+	if req.ProcessStartConfig.IdReusePolicy == nil {
+		req.ProcessStartConfig.IdReusePolicy = xdbapi.ALLOW_IF_NO_RUNNING.Ptr()
```
Comment (longquanzheng) [inline persistence/sql/sql_process_store.go:155]: this is dup code as three lines above.

File: persistence/sql/sql_process_store.go [hunk 10]
```diff
@@ -91,28 +92,372 @@ func (p sqlProcessStoreImpl) StopProcess(ctx context.Context, request persistenc
...
+	}
+
+	switch *req.ProcessStartConfig.IdReusePolicy {
```
Comment (longquanzheng) [inline persistence/sql/sql_process_store.go:158]: nice! I like how it's organized here 👍

File: persistence/sql/sql_process_store.go [hunk 11]
```diff
@@ -91,28 +92,372 @@ func (p sqlProcessStoreImpl) StopProcess(ctx context.Context, request persistenc
...
+		}
+
+		return &persistence.StartProcessResponse{
+			ProcessExecutionId: prcExeId,
+			AlreadyStarted:     false,
+			HasNewWorkerTask:   hasNewWorkerTask,
+		}, nil
+	}
+
+	prcExeId := uuid.MustNewUUID()
```
Comment (longquanzheng) [inline persistence/sql/sql_process_store.go:259]: nit: the below code can be abstracted in a separate method for reuse:
```
p.insertNewLatestProcessExecution(...)
```

File: persistence/sql/sql_process_store.go [hunk 12]
```diff
@@ -91,28 +92,372 @@ func (p sqlProcessStoreImpl) StopProcess(ctx context.Context, request persistenc
...
 			}, nil
 		}
+
+		// if it is not running, but completed normally, return error
+		// otherwise, start a new process
+		if processExecutionRowForUpdate.Status == persistence.ProcessExecutionStatusCompleted {
+			return nil, fmt.Errorf("process %v has completed normally", request.Request.ProcessId)
+		}
+
+		prcExeId := uuid.MustNewUUID()
```
Comment (longquanzheng) [inline persistence/sql/sql_process_store.go:313]: similar here `p.UpdateLatestAndInsertNewProcessExecution`

File: persistence/sql/sqltest/base_basic.go [hunk 13]
```diff
@@ -118,6 +118,245 @@ func SQLBasicTest(ass *assert.Assertions, store persistence.ProcessStore) {
...
+	// Start the process and verify it started correctly.
+	prcExeId := startProcess(ctx, ass, store, namespace, processId, input)
+
+	// Try to start the process again and verify the behavior.
+	retryStartProcessForFailure(ctx, ass, store, namespace, processId, input)
+
+	// Describe the process.
+	describeProcess(ctx, ass, store, namespace, processId, xdbapi.RUNNING)
+
+	// Test waitUntil API execution
```
Comment (longquanzheng) [inline persistence/sql/sqltest/base_basic.go:181]: you probably don't have to test the worker task code path. I like the way the above test cases doing it to be more straightfoward

File: persistence/sql/sqltest/base_basic.go [hunk 14]
```diff
@@ -118,6 +118,245 @@ func SQLBasicTest(ass *assert.Assertions, store persistence.ProcessStore) {
 	describeProcess(ctx, ass, store, namespace, processId, xdbapi.COMPLETED)
 }
 
+func SQLProcessIdReusePolicyAllowIfPreviousExitAbnormally(ass *assert.Assertions, store persistence.ProcessStore) {
```
Comment (longquanzheng) [inline persistence/sql/sqltest/base_basic.go:121]: very clean and clear test case 👍

File: persistence/sql/sqltest/common.go [hunk 15]
```diff
@@ -59,6 +59,150 @@ func startProcess(
...
+		Namespace:       namespace,
+		ProcessId:       processId,
+		ProcessStopType: xdbapi.TERMINATE,
+	})
+
+	ass.Nil(err)
+	ass.False(resp.NotExists)
+}
+
+func startProcessWithAllowIfPreviousExitAbnormally(ctx context.Context, ass *assert.Assertions,
```
Comment (longquanzheng) [inline persistence/sql/sqltest/common.go:74]: It seems like we are missing a test case for the default policy. Maybe add it into the test cases like `SQLProcessIdReusePolicyTerminateIfRunning` to see if it return error with the default policy

**Comment by longquanzheng** [comment]
Very good job on the code quality/readability!

File: persistence/sql/sql_process_store.go [hunk 16]
```diff
@@ -16,6 +16,8 @@ package sql
 import (
 	"context"
 	"fmt"
+	"time"
```
Comment (longquanzheng) [inline persistence/sql/sql_process_store.go:19]: this file is getting big. I think we can consider to split it based on API (you can do it in separate PR):
* sql_store_start_process.go
* sql_store_stop_process.go
* sql_store_common.go
cc @zklgame

File: extensions/postgres/transactional.go (hunk 7 above)
Comment (duoertai) [reply extensions/postgres/transactional.go:41]: right, here I use selectContext is because there could be 0 or 1 row is returned, getContext will consider 0 row is an error, thus abort the transaction

File: persistence/sql/sqltest/base_basic.go (hunk 13 above)
Comment (duoertai) [reply persistence/sql/sqltest/base_basic.go:181]: this is to make the process complete and then start again

File: extensions/postgres/transactional.go (hunk 7 above)
Comment (longquanzheng) [reply extensions/postgres/transactional.go:41]: I see. It’s probably better to return (Row, bool, error) and use the bool to indicate empty.

File: persistence/sql/sqltest/base_basic.go (hunk 13 above)
Comment (longquanzheng) [reply persistence/sql/sqltest/base_basic.go:181]: Ohh I see . Makes sense to me

File: persistence/sql/sql_process_store.go (hunk 16 above)
Comment (duoertai) [reply persistence/sql/sql_process_store.go:19]: sure, I'll do it in the next PR

File: persistence/sql/sql_process_store.go [hunk 17]
```diff
@@ -91,28 +92,372 @@ func (p sqlProcessStoreImpl) StopProcess(ctx context.Context, request persistenc
...
-		if p.session.IsDupEntryError(err) {
-			// TODO support other ProcessIdReusePolicy on this error
+		return nil, err
+	}
+	if len(latestProcessExecutions) > 1 {
+		return nil, fmt.Errorf("more than one running process execution for process id %v", request.Request.ProcessId)
+	}
+
+	if len(latestProcessExecutions) == 1 {
+		processExecutionRowForUpdate, err := tx.SelectProcessExecutionForUpdate(ctx, latestProcessExecutions[0].ProcessExecutionId)
```
[the remark this answers is not in the document]
Comment (longquanzheng) [reply persistence/sql/sql_process_store.go]: We need to add a other `SelectProcessExecution` (no ForUpdate) for ALLOW_IF_NO_RUNNING, ALLOW_IF_PREIVOUS_EXIT_ABNOARMALLY because they are not updating the current execution. This will save the locking contention.

**Review by longquanzheng — APPROVED** [review]
https://github.com/xdblab/xdb/pull/36#discussion_r1361431465
https://github.com/xdblab/xdb/pull/36#discussion_r1361419744 

lgtm, with two last comments ^
